tests/unit/test_charm_configure.py

```diff
@@ -0,0 +1,753 @@
+            )
+            nrf_relation = scenario.Relation(
+                endpoint="fiveg_nrf",
+                interface="fiveg-nrf",
+                remote_app_data={"url": TEST_NRF_URL},
+            )
+            nms_relation = scenario.Relation(
+                endpoint="sdcore_config",
+                interface="sdcore-config",
+                remote_app_data={"webui_url": "whatever"},
+            )
+            certs_mount = scenario.Mount(
+                location="/support/TLS",
+                src=tempdir,
+            )
+            config_mount = scenario.Mount(
+                location="/free5gc/config",
+                src=tempdir,
+            )
+            container = scenario.Container(
+                name=CONTAINER_NAME,
+                can_connect=True,
+                mounts={"certs": certs_mount, "config": config_mount},
+            )
+            state_in = scenario.State(
+                containers=[container],
+                relations=[certificates_relation, nrf_relation, nms_relation],
+                leader=True,
+            )
+            self.mock_check_output.return_value = TEST_POD_IP
+            self.mock_generate_csr.return_value = TEST_CSR
+
+            with open(f"{tempdir}/ausf.key", "w") as f:
+                f.write(TEST_PRIVATE_KEY.decode())
+
+            self.ctx.run("update_status", state_in)
+
+            self.mock_request_certificate_creation.assert_called_once()
+
+    def test_given_charm_workload_is_ready_to_configure_and_provider_certificate_needs_updating_when_update_status_then_new_provider_certificate_is_pushed_to_the_container(  # noqa: E501
+        self,
+    ):
+        with tempfile.TemporaryDirectory() as tempdir:
+            certificates_relation = scenario.Relation(
+                endpoint="certificates",
+                interface="tls-certificates",
+            )
+            nrf_relation = scenario.Relation(
+                endpoint="fiveg_nrf",
+                interface="fiveg-nrf",
+                remote_app_data={"url": TEST_NRF_URL},
+            )
+            nms_relation = scenario.Relation(
+                endpoint="sdcore_config",
+                interface="sdcore-config",
+                remote_app_data={"webui_url": "whatever"},
+            )
+            certs_mount = scenario.Mount(
+                location="/support/TLS",
+                src=tempdir,
+            )
+            config_mount = scenario.Mount(
+                location="/free5gc/config",
+                src=tempdir,
+            )
+            container = scenario.Container(
+                name=CONTAINER_NAME,
+                can_connect=True,
+                mounts={"certs": certs_mount, "config": config_mount},
+            )
+            state_in = scenario.State(
+                containers=[container],
+                relations=[certificates_relation, nrf_relation, nms_relation],
+                leader=True,
+            )
```

Comment:
How about we extract all that to a function returning `state_in` instead of defining the same code block in each test?

Reply:
I also think about it. Some inputs could change from test to test, however many of them are repeated code blocks. I support the idea to create functions to decrease the repetiition.

Reply:
I'm not sure how much I agree here, as it stands, a unit test is self explanatory and contains everything you need to know. If we extract state_in, my fear is that we loose test independence and readability. In other words, I don't really mind the duplication.

Reply:
We loose the independence, but it was never required in this part of the tests - they all assume that the starting point is a workload ready for configuration. I'm not sure about readability, but even if we do loose some of it (not much IMO) isn't improved maintainability a bigger value?

Reply:
Creating some helpers functions gets  necessary input and returns new objects every time does not cause affect the independence and reduces some lines of code.  It could stay like this. Not a big issue.